pkg/controller/tls/letsencrypt.go

```diff
@@ -560,6 +563,44 @@ func (u *LEUser) httpChallenge(ctx context.Context, domain string) (*certificate
 		return nil, err
 	}
 
+	// Setup and query test acme-challenge endpoint. This will ensure the networking from ingress to server is working
+	// before attempting to request the cert
+	err = httpProviderServer.Present(domain, "test", "test")
+	if err != nil {
+		return nil, fmt.Errorf("failed to present test acme-challenge. %w", err)
+	}
+
+	err = wait.ExponentialBackoffWithContext(ctx, wait.Backoff{
+		Duration: 1 * time.Second,
+		Factor:   2,
+		Steps:    10,
+		Cap:      180 * time.Second,
+	}, func(ctx context.Context) (done bool, err error) {
+		testUrl := "http://" + domain + acmeChallengePath + "test"
+		logrus.Debugf("Attempting to hit test cert path %v", testUrl)
+		resp, err := http.Get(testUrl)
+		if err != nil {
+			logrus.Errorf("Error hitting test acme-challenge endpoint: %v", err)
+			return false, nil
+		}
+		logrus.Debugf("Response code for test acme-challenge endpoint: %v", resp.StatusCode)
+		if resp.StatusCode == 200 {
+			return true, nil
+		}
+		return false, nil
+	})
+
+	// The provider will only present a single token at a time. Cleanup before attempting to obtain cert.
+	// Need to do this before potentially returning from ExponentialBackoffWithContext's error
+	cleanupErr := httpProviderServer.CleanUp(domain, "test", "test")
+	if cleanupErr != nil {
+		logrus.Errorf("Failed to cleanup test acme-challenge. %v", cleanupErr)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to successfully hit test acme-challenge endpoint. %w", err)
+	}
```

Comment:
nit: I think factoring this out to helper function would clean things up a bit:

```go
func ensureChallengeEndpoint(ctx context.Context, provider *http01.ProviderServer, domain string) error {
    if err := provider.Present(domain, "test", "test"); err != nil {
        return nil, fmt.Errorf("failed to present test acme-challenge. %w", err)
    }
    
    defer func() {
        // The provider will only present a single token at a time. Cleanup before attempting to obtain cert.
        // Need to do this before potentially returning from ExponentialBackoffWithContext's error
        if err := httpProviderServer.CleanUp(domain, "test", "test"); err != nil {
            logrus.Errorf("Failed to cleanup test acme-challenge. %v", err)     
        }
    }()
    
    url := "http://" + domain + acmeChallengePath + "test"
	if err := wait.ExponentialBackoffWithContext(ctx, wait.Backoff{
        Duration: 1 * time.Second,
		Factor:   2,
		Steps:    10,
		Cap:      180 * time.Second,
	}, func(ctx context.Context) (bool, error) {
		logrus.Debugf("Attempting to hit test cert path %v", url)
		resp, err := http.Get(url)
		if err != nil {
			logrus.Errorf("Error hitting test acme-challenge endpoint: %v", err)
			return false, nil
		}
		logrus.Debugf("Response code for test acme-challenge endpoint: %v", resp.StatusCode)
		if resp.StatusCode == 200 {
			return true, nil
		}
		return false, nil
	}); err != nil {
        return fmt.Errorf("failed to successfully hit test acme-challenge endpoint. %w", err)
	}
	
	return nil
}
```

Reply:
i did the refactor